src/EFCore.PG/Migrations/Internal/NpgsqlMigrator.cs

```diff
@@ -0,0 +1,139 @@
+
+namespace Npgsql.EntityFrameworkCore.PostgreSQL.Migrations.Internal;
+
+// Migrator is EF-pubternal, but overriding it is the only way to force Npgsql to ReloadTypes() after executing a migration which adds
+// types to the database
+#pragma warning disable EF1001
+
+/// <summary>
+///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+///     any release. You should only use it directly in your code with extreme caution and knowing that
+///     doing so can result in application failures when updating to a new Entity Framework Core release.
+/// </summary>
+public class NpgsqlMigrator : Migrator
+{
+    private readonly IHistoryRepository _historyRepository;
+    private readonly IRelationalConnection _connection;
+
+    /// <summary>
+    ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+    ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+    ///     any release. You should only use it directly in your code with extreme caution and knowing that
+    ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+    /// </summary>
+    public NpgsqlMigrator(
+        IMigrationsAssembly migrationsAssembly,
+        IHistoryRepository historyRepository,
+        IDatabaseCreator databaseCreator,
+        IMigrationsSqlGenerator migrationsSqlGenerator,
+        IRawSqlCommandBuilder rawSqlCommandBuilder,
+        IMigrationCommandExecutor migrationCommandExecutor,
+        IRelationalConnection connection,
+        ISqlGenerationHelper sqlGenerationHelper,
+        ICurrentDbContext currentContext,
+        IModelRuntimeInitializer modelRuntimeInitializer,
+        IDiagnosticsLogger<DbLoggerCategory.Migrations> logger,
+        IRelationalCommandDiagnosticsLogger commandLogger,
+        IDatabaseProvider databaseProvider)
+        : base(migrationsAssembly, historyRepository, databaseCreator, migrationsSqlGenerator, rawSqlCommandBuilder,
+            migrationCommandExecutor, connection, sqlGenerationHelper, currentContext, modelRuntimeInitializer, logger,
+            commandLogger, databaseProvider)
+    {
+        _historyRepository = historyRepository;
+        _connection = connection;
+    }
+
+    /// <summary>
+    ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+    ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+    ///     any release. You should only use it directly in your code with extreme caution and knowing that
+    ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+    /// </summary>
+    public override void Migrate(string? targetMigration = null)
```

Comment:
@bricelam look at this lovely thing... I refused to override Migrator for years (because it's internal) but I finally did it.

Basically certain AlterDatabaseOperations create new PG types in the database (e.g. when you add an extension), and then if you're applying migrations within the process with Migrate(), you need to tell Npgsql to reload the types at the ADO.NET layer. This has been implemented cleanly for a long time in NpgsqlDatabaseCreator (so EnsuredCreated worked), but not Migrate().

Ideally there would be some hooks after the migrations are applied, which would be called both from EnsureCreated() and Migrate(); they'd need to receive the migration operations in order to inspect them etc. Here I have to calculate the applied migrations and the ones to be applied (so once here just for this reload, once in Migrator for real).

Reply:
Sounds like a good scenario we should cover as part of https://github.com/dotnet/efcore/issues/24710